File: app-shared/Sources/CommonLibraryCore/Business/ConfigManager.swift
// SPDX-FileCopyrightText: 2026 Davide De Rosa
//
// SPDX-License-Identifier: GPL-3.0

import Dispatch

#if !PSP_CROSS
extension ConfigManager: ObservableObject {}
#endif

@MainActor
public final class ConfigManager {
    private let queue = DispatchQueue(label: "ConfigManager")

    private let strategy: ConfigManagerStrategy?

    private let buildNumber: Int

    private var bundle: ABI.ConfigBundle? {
        willSet {
#if !PSP_CROSS
            objectWillChange.send()
#endif
        }
        didSet {
            // Take a snapshot of the active flags
            queue.sync {
                synchronousActiveFlags = Set(ABI.ConfigFlag.allCases.filter {
                    isActive($0)
                })
            }
        }
    }

    nonisolated(unsafe)
    private var synchronousActiveFlags: Set<ABI.ConfigFlag> = []

    private var isPending = false

    public init() {
        strategy = nil
        buildNumber = .max // Activate flags regardless of .minBuild
    }

    public init(strategy: ConfigManagerStrategy, buildNumber: Int) {
        self.strategy = strategy
        self.buildNumber = buildNumber
    }

    // TODO: #1447, handle 0-100 deployment values with local random value
    public func refreshBundle() async {
        guard let strategy else {
            return
        }
        guard !isPending else {
            return
        }
        isPending = true
        defer {
            isPending = false
        }
        do {
            pp_log_g(.App.core, .debug, "Config: refreshing bundle...")
            let newBundle = try await strategy.bundle()
            bundle = newBundle
            let activeFlags = newBundle.activeFlags(withBuild: buildNumber)
            pp_log_g(.App.core, .info, "Config: active flags = \(activeFlags)")
            pp_log_g(.App.core, .debug, "Config: \(newBundle)")
        } catch ABI.AppError.rateLimit {
            pp_log_g(.App.core, .debug, "Config: TTL")
        } catch {
            pp_log_g(.App.core, .error, "Unable to refresh config flags: \(error)")
        }
    }

    public func isActive(_ flag: ABI.ConfigFlag) -> Bool {
        activeMap(for: flag) != nil
    }

    public func data(for flag: ABI.ConfigFlag) -> JSON? {
        activeMap(for: flag)?.data
    }

    public nonisolated var activeFlags: Set<ABI.ConfigFlag> {
        queue.sync {
            synchronousActiveFlags
        }
    }
}

private extension ConfigManager {
    func activeMap(for flag: ABI.ConfigFlag) -> ABI.ConfigBundle.Config? {
        guard let map = bundle?.map[flag] else {
            return nil
        }
        guard map.isActive(withBuild: buildNumber) else {
            return nil
        }
        return map
    }
}

File: app-android/app/src/main/kotlin/com/algoritmico/passepartout/business/managers/ConfigManager.kt
// SPDX-FileCopyrightText: 2026 Davide De Rosa
//
// SPDX-License-Identifier: GPL-3.0

package com.algoritmico.passepartout.business.managers

import com.algoritmico.passepartout.context.AppLog
import com.algoritmico.passepartout.business.extensions.JSON
import com.algoritmico.passepartout.business.extensions.runCatchingNonFatal
import com.algoritmico.passepartout.context.newEventFlow
import com.algoritmico.passepartout.models.ConfigBundleConfig
import com.algoritmico.passepartout.models.ConfigEventRefresh
import com.algoritmico.passepartout.models.ConfigFlag
import com.algoritmico.passepartout.models.Event
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

interface ConfigManagerStrategy {
    suspend fun bundle(): ConfigBundle
    fun resetTTL()
}

sealed class ConfigManagerException: Exception() {
    data object RateLimit: ConfigManagerException()
}

class ConfigManager(
    private val logTag: String,
    private val strategy: ConfigManagerStrategy,
    private val buildNumber: Int = Int.MAX_VALUE
) {
    private val refreshMutex = Mutex()
    private val bundleLock = Any()
    private var bundle: ConfigBundle? = null

    private val _events = newEventFlow()
    val events: SharedFlow<Event> = _events.asSharedFlow()

    suspend fun refreshBundle(): Boolean {
        if (!refreshMutex.tryLock()) {
            return false
        }
        return runCatchingNonFatal {
            AppLog.d(logTag, "Config: refreshing bundle...")
            val newBundle = strategy.bundle()
            val event = synchronized(bundleLock) {
                bundle = newBundle
                refreshEvent(newBundle)
            }
            _events.emit(event)
            AppLog.i(logTag, "Config: active flags = ${event.flags}")
            AppLog.d(logTag, "Config: $newBundle")
            true
        }.also {
            refreshMutex.unlock()
        }.getOrElse {
            when (it) {
                is ConfigManagerException.RateLimit -> {
                    AppLog.d(logTag, "Config: TTL")
                    false
                }
                else -> {
                    AppLog.e(logTag, "Config: Unable to refresh flags", it)
                    throw it
                }
            }
        }
    }

    fun resetTTL() {
        strategy.resetTTL()
    }

    fun isActive(flag: ConfigFlag): Boolean {
        return activeConfig(flag) != null
    }

    fun data(flag: ConfigFlag): JsonElement? {
        return activeConfig(flag)?.data
    }

    val activeFlags: Set<ConfigFlag>
        get() = synchronized(bundleLock) {
            bundle?.activeFlags(buildNumber) ?: emptySet()
        }

    private fun activeConfig(flag: ConfigFlag): ConfigBundleConfig? {
        return synchronized(bundleLock) {
            bundle?.map?.get(flag)?.takeIf {
                it.isActive(buildNumber)
            }
        }
    }

    private fun refreshEvent(bundle: ConfigBundle): ConfigEventRefresh {
        val activeFlags = bundle.activeFlags(buildNumber)
            .sortedBy { it.ordinal }
        val data = activeFlags
            .mapNotNull { flag ->
                bundle.map[flag]?.data?.let { flag.value to it }
            }
            .toMap()
        return ConfigEventRefresh(
            flags = activeFlags,
            data = JsonObject(data)
        )
    }
}

class ConfigBundle(
    val map: Map<ConfigFlag, ConfigBundleConfig>
) {
    fun activeFlags(buildNumber: Int): Set<ConfigFlag> {
        return map
            .filterValues { it.isActive(buildNumber) }
            .keys
    }

    override fun toString(): String {
        return "ConfigBundle(map=$map)"
    }

    companion object {
        fun decode(data: ByteArray): ConfigBundle {
            val map = JSON
                .decode<Map<String, ConfigBundleConfig>>(data.decodeToString())
                .mapNotNull { (key, value) ->
                    ConfigFlag.decode(key)?.let { it to value }
                }
                .toMap()
            return ConfigBundle(map)
        }
    }
}

private fun ConfigBundleConfig.isActive(buildNumber: Int): Boolean {
    if (minBuild != null && buildNumber < minBuild) {
        return false
    }
    return rate == 100
}
